src/cdk-experimental/scrolling/auto-size-virtual-scroll.ts

```diff
@@ -0,0 +1,228 @@
+   * @param averager The averager used to estimate the size of unseen items.
+   */
+  constructor(minBufferPx: number, addBufferPx: number, averager = new ItemSizeAverager()) {
+    this._minBufferPx = minBufferPx;
+    this._addBufferPx = addBufferPx;
+    this._averager = averager;
+  }
+
+  /**
+   * Attaches this scroll strategy to a viewport.
+   * @param viewport The viewport to attach this strategy to.
+   */
+  attach(viewport: CdkVirtualScrollViewport) {
+    this._viewport = viewport;
+    this._updateTotalContentSize();
+    this._renderContentForOffset(this._viewport.measureScrollOffset());
+  }
+
+  /** Detaches this scroll strategy from the currently attached viewport. */
+  detach() {
+    this._viewport = null;
+  }
+
+  /** Called when the viewport is scrolled. */
+  onContentScrolled() {
+    if (this._viewport) {
+      this._renderContentForOffset(this._viewport.measureScrollOffset());
+    }
+  }
+
+  /** Called when the length of the data changes. */
+  onDataLengthChanged() {
+    if (this._viewport) {
+      this._updateTotalContentSize();
+      this._renderContentForOffset(this._viewport.measureScrollOffset());
+    }
+  }
+
+  /**
+   * Update the buffer parameters.
+   * @param minBufferPx The minimum amount of buffer rendered beyond the viewport (in pixels).
+   * @param addBufferPx The number of buffer items to render beyond the edge of the viewport (in
+   *     pixels).
+   */
+  updateBufferSize(minBufferPx, addBufferPx) {
+    this._minBufferPx = minBufferPx;
+    this._addBufferPx = addBufferPx;
+  }
+
+  /**
+   * Render the content that we estimate should be shown for the given scroll offset.
+   * Note: must not be called if `this._viewport` is null
+   */
+  private _renderContentForOffset(scrollOffset: number) {
+    const viewport = this._viewport!;
+    const itemSize = this._averager.getAverageItemSize();
+    const firstVisibleIndex =
+        Math.min(viewport.getDataLength() - 1, Math.floor(scrollOffset / itemSize));
+    const bufferSize = Math.ceil(this._addBufferPx / itemSize);
+    const range = this._expandRange(
+        this._getVisibleRangeForIndex(firstVisibleIndex), bufferSize, bufferSize);
+
+    viewport.setRenderedRange(range);
+    viewport.setRenderedContentOffset(itemSize * range.start);
+  }
+
+  // TODO: maybe move to base class, can probably share with fixed size strategy.
+  /**
+   * Gets the visible range of data for the given start index. If the start index is too close to
+   * the end of the list it may be backed up to ensure the estimated size of the range is enough to
+   * fill the viewport.
+   * Note: must not be called if `this._viewport` is null
+   * @param startIndex The index to start the range at
+   * @return a range estimated to be large enough to fill the viewport when rendered.
+   */
+  private _getVisibleRangeForIndex(startIndex: number): Range {
+    const viewport = this._viewport!;
+    let range = {
```

Comment:
Add a type to this one so it breaks if we change the `Range` interface. Also you should be able to turn it into a const.